contracts/AelinUpFrontDeal.sol

```diff
@@ -257,43 +262,48 @@ contract AelinUpFrontDeal is AelinERC20, MinimalProxyFactory, IAelinUpFrontDeal
             uint256 _underlyingDealTokenTotal = dealConfig.underlyingDealTokenTotal;
             // Claim Deal Tokens
             bool deallocate = totalPoolShares > _underlyingDealTokenTotal;
+            uint256 adjustedShareAmountForUser;
+            uint256 precisionAdjustedRefund;
 
             if (deallocate) {
                 // adjust for deallocation and mint deal tokens
-                uint256 adjustedDealTokensForUser = (((poolSharesPerUser[msg.sender] * _underlyingDealTokenTotal) /
-                    totalPoolShares) * (BASE - AELIN_FEE - dealData.sponsorFee)) / BASE;
+                adjustedShareAmountForUser =
+                    (((poolSharesPerUser[msg.sender] * _underlyingDealTokenTotal) / totalPoolShares) *
+                        (BASE - AELIN_FEE - dealData.sponsorFee)) /
+                    BASE;
                 poolSharesPerUser[msg.sender] = 0;
 
                 // refund any purchase tokens that got deallocated
                 uint256 purchasingRefund = purchaseTokensPerUser[msg.sender] -
                     ((purchaseTokensPerUser[msg.sender] * _underlyingDealTokenTotal) / totalPoolShares);
-                purchaseTokensPerUser[msg.sender] = 0;
 
-                uint256 precisionAdjustedRefund = purchasingRefund > IERC20(_purchaseToken).balanceOf(address(this))
+                precisionAdjustedRefund = purchasingRefund > IERC20(_purchaseToken).balanceOf(address(this))
                     ? IERC20(_purchaseToken).balanceOf(address(this))
                     : purchasingRefund;
 
-                // mint deal tokens and transfer purchase token refund
-                _mint(msg.sender, adjustedDealTokensForUser);
+                // Transfer purchase token refund
                 IERC20(_purchaseToken).safeTransfer(msg.sender, precisionAdjustedRefund);
-
-                emit ClaimDealTokens(msg.sender, adjustedDealTokensForUser, precisionAdjustedRefund);
             } else {
                 // mint deal tokens when there is no deallocation
-                uint256 adjustedDealTokensForUser = ((BASE - AELIN_FEE - dealData.sponsorFee) *
-                    poolSharesPerUser[msg.sender]) / BASE;
-                poolSharesPerUser[msg.sender] = 0;
-                purchaseTokensPerUser[msg.sender] = 0;
-                _mint(msg.sender, adjustedDealTokensForUser);
-                emit ClaimDealTokens(msg.sender, adjustedDealTokensForUser, 0);
+                adjustedShareAmountForUser =
+                    ((BASE - AELIN_FEE - dealData.sponsorFee) * poolSharesPerUser[msg.sender]) /
+                    BASE;
             }
+            poolSharesPerUser[msg.sender] = 0;
+            purchaseTokensPerUser[msg.sender] = 0;
+
+            // mint vesting token
+            _mint(msg.sender, tokenCount);
+            tokenDetails[tokenCount] = TokenDetails(adjustedShareAmountForUser, block.timestamp);
```

Comment:
this block.timestamp feels wrong to me although I am about to look at the claiming changes. this is not the lastClaimedAt timestamp since this is just when they settle their allocations. In fact the user should only have to call purchaser claim when deallocations are allowed if that makes sense. this should be re-architected a bit in the future

Reply:
they are not claiming at all here so I assume setting the lastClaimedAt will cause a bug with vesting maybe. will need to check. we should definitely add a test where a user calls purchaserClaim halfway through the vesting schedule

Reply:
I think the correct solution here is to do this 
tokenDetails[tokenCount] = TokenDetails(adjustedShareAmountForUser, purchaseExpiry);

using purchaseExpiry means that the lastClaimedAt for each user will start at the beginning of the vesting schedule

Reply:
It's a really good find even though it's safe to put `block.timestamp`. Have a look at `claimableUnderlyingTokens`

```
  uint256 minTime = schedule.lastClaimedAt > vestingCliffExpiry ? schedule.lastClaimedAt : vestingCliffExpiry;
 ```
 
 When accepting a deal we always assume `block.timestamp < purchaseExpiry` and that `purchaseExpiry <= vestingCliffExpiry`

As you can see in the code above, `minTime` will only use `lastClaimedAt` if it's greater than `vestingCliffExpiry`

So it's safe.